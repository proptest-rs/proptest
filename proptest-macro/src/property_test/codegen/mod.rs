use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::{parse_str, spanned::Spanned, Attribute, Ident, ItemFn, PatType};

use super::{options::Options, utils::strip_args};

mod test_body;

/// Generate the modified test function
///
/// The rough process is:
///  - strip out the function args from the provided function
///  - turn them into a struct
///  - implement `Arbitrary` for that struct (simple field-wise impl)
///  - create a runner, do the rest
///
///  Currently, any attributes on parameters are ignored - in the future, we probably want to read
///  these for things like customizing strategies
pub(super) fn generate(item_fn: ItemFn, options: Options) -> TokenStream {
    let (mut argless_fn, args) = strip_args(item_fn);

    let struct_tokens = generate_struct(&argless_fn.sig.ident, &args);
    let arb_tokens = generate_arbitrary_impl(&argless_fn.sig.ident, &args);

    let struct_and_tokens = quote! {
        #struct_tokens
        #arb_tokens
    };

    let new_body = test_body::body(
        *argless_fn.block,
        &args,
        struct_and_tokens,
        &argless_fn.sig.ident,
        &argless_fn.sig.output,
        &options,
    );

    *argless_fn.block = new_body;
    argless_fn.attrs.push(test_attr());

    argless_fn.to_token_stream()
}

/// Generate the inner struct that represents the arguments of the function
fn generate_struct(fn_name: &Ident, args: &[PatType]) -> TokenStream {
    let struct_name = struct_name(fn_name);

    let fields = args.iter().enumerate().map(|(index, arg)| {
        let field_name = nth_field_name(&arg.pat, index);
        let ty = &arg.ty;

        quote! { #field_name: #ty, }
    });

    quote! {
        #[derive(Debug)]
        struct #struct_name {
            #(#fields)*
        }
    }
}

/// Generate the arbitrary impl for the struct
fn generate_arbitrary_impl(fn_name: &Ident, args: &[PatType]) -> TokenStream {
    let struct_name = struct_name(fn_name);

    let arg_types = args.iter().map(|arg| {
        let ty = &arg.ty;
        quote!(#ty,)
    });

    let arg_types = quote! { #(#arg_types)* };

    let arg_names = args.iter().enumerate().map(|(index, ty)| {
        let name = nth_field_name(ty.span(), index);
        quote!(#name,)
    });

    let arg_names = quote! { #(#arg_names)* };

    quote! {
        impl ::proptest::prelude::Arbitrary for #struct_name {
            type Parameters = ();
            type Strategy = ::proptest::strategy::Map<::proptest::arbitrary::StrategyFor<(#arg_types)>, fn((#arg_types)) -> Self>;

            fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
                use ::proptest::strategy::Strategy;
                ::proptest::prelude::any::<(#arg_types)>().prop_map(|(#arg_names)| Self { #arg_names })
            }
        }
    }
}

/// Convert the name of a function to the name of a struct representing its args
///
/// E.g. `some_function` -> `SomeFunctionArgs`
fn struct_name(fn_name: &Ident) -> Ident {
    use convert_case::{Case, Casing};

    let name = fn_name.to_string();
    let name = name.to_case(Case::Pascal);
    let name = format!("{name}Args");
    Ident::new(&name, fn_name.span())
}

/// We convert all fields to `"field0"`, etc. to account for various different patterns that can
/// exist in function args. We restore the patterns/bindings when we destructure the struct in the
/// test body
fn nth_field_name(span: impl Spanned, index: usize) -> Ident {
    Ident::new(&format!("field{index}"), span.span())
}

/// I couldn't find a better way to get just the `#[test]` attribute since [`syn::Attribute`]
/// doesn't implement `Parse`
fn test_attr() -> Attribute {
    let mut f: ItemFn = parse_str("#[test] fn foo() {}").unwrap();
    f.attrs.pop().unwrap()
}

#[cfg(test)]
mod tests {
    use quote::ToTokens;
    use syn::{parse2, parse_str, ItemStruct};

    use super::*;

    /// Simple helper that parses a function, and validates that the struct name and fields are
    /// correct
    fn check_struct(
        fn_def: &str,
        expected_name: &'static str,
        expected_fields: impl IntoIterator<Item = (&'static str, &'static str)>,
    ) {
        let f: ItemFn = parse_str(fn_def).unwrap();
        let (f, args) = strip_args(f);
        let tokens = generate_struct(&f.sig.ident, &args);
        let s: ItemStruct = parse2(tokens).unwrap();

        let fields: Vec<_> = s
            .fields
            .into_iter()
            .map(|field| {
                (
                    field.ident.unwrap().to_string(),
                    field.ty.to_token_stream().to_string(),
                )
            })
            .collect();

        assert_eq!(s.ident.to_string(), expected_name);
        let expected_fields: Vec<_> = expected_fields
            .into_iter()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect();

        assert_eq!(fields, expected_fields);
    }

    #[test]
    fn derives_debug() {
        let f: ItemFn = parse_str("fn foo(x: i32) {}").unwrap();
        let (f, args) = strip_args(f);
        let string = generate_struct(&f.sig.ident, &args).to_string();

        assert!(string.contains("derive"));
        assert!(string.contains("Debug"));
    }

    #[test]
    fn generates_correct_struct() {
        check_struct("fn foo() {}", "FooArgs", []);
        check_struct("fn foo(x: i32) {}", "FooArgs", [("field0", "i32")]);
        check_struct(
            "fn foo(a: i32, b: String) {}",
            "FooArgs",
            [("field0", "i32"), ("field1", "String")],
        );
    }

    #[test]
    fn generates_arbitrary_impl() {
        let f: ItemFn = parse_str("fn foo(x: i32, y: u8) {}").unwrap();
        let (f, args) = strip_args(f);
        let arb = generate_arbitrary_impl(&f.sig.ident, &args);

        let expected = quote! {
            impl ::proptest::prelude::Arbitrary for FooArgs {
                type Parameters = ();
                type Strategy = ::proptest::strategy::Map<::proptest::arbitrary::StrategyFor<(i32, u8,)>, fn((i32, u8,)) -> Self>;

                fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
                    use ::proptest::strategy::Strategy;

                    ::proptest::prelude::any::<(i32, u8,)>().prop_map(|(field0, field1,)| Self { field0, field1, })
                }

            }
        };

        assert_eq!(arb.to_string(), expected.to_string());
    }
}

#[cfg(test)]
mod snapshot_tests {
    use super::*;

    macro_rules! snapshot_test {
        ($name:ident) => {
            #[test]
            fn $name() {
                const TEXT: &str = include_str!(concat!(
                    "test_data/",
                    stringify!($name),
                    ".rs"
                ));

                let tokens = generate(
                    parse_str(TEXT).unwrap(),
                    $crate::property_test::options::Options::default(),
                );
                insta::assert_debug_snapshot!(tokens);
            }
        };
    }

    snapshot_test!(simple);
    snapshot_test!(many_params);
    snapshot_test!(arg_pattern);
}
