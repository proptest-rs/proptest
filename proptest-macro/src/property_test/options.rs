use proc_macro2::TokenStream;
use quote::quote_spanned;
use syn::{
    parse::Parse, punctuated::Punctuated, spanned::Spanned, Expr, Ident,
    LitStr, MetaNameValue, Token,
};

/// Options parsed from the attribute itself (e.g. the config from `#[property_test(config = ...)]`)
#[derive(Default)]
pub(super) struct Options {
    /// Collect compiler errors and emit them later, since errors here are largely recoverable
    pub errors: Vec<TokenStream>,
    pub config: Option<Expr>,
}

impl Parse for Options {
    // note: this impl takes only the contents of the attr, not the attr itself
    // e.g. it will get `foo = bar, baz = qux`, not `#[macro(foo = bar, baz = qux)]`
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let pairs =
            Punctuated::<MetaNameValue, Token![,]>::parse_terminated(input)?;

        let mut errors = Vec::new();

        let mut config = None;

        for MetaNameValue { path, value, .. } in pairs {
            let path_string = path.get_ident().map(Ident::to_string);

            match path_string.as_deref() {
                None => errors.push(quote_spanned!(path.span() => compile_error!("unknown argument"))),
                Some("config") => config = Some(value),
                Some(other) => {
                    let error_message = format!("unknown argument: {other}");
                    let error_message = LitStr::new(&error_message, other.span());
                    let error = quote_spanned!(other.span() => compile_error!(#error_message));
                    errors.push(error);
                }
            }
        }

        Ok(Self { errors, config })
    }
}

#[cfg(test)]
mod tests {
    use syn::parse_str;

    use super::*;

    #[test]
    fn simple_parse_example() {
        let Options { errors, config } =
            parse_str("config = (), random = 123").unwrap();

        assert!(config.is_some());
        assert_eq!(errors.len(), 1);
    }
}
