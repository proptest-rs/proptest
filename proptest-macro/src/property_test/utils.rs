use core::mem::replace;

use syn::{punctuated::Punctuated, FnArg, ItemFn, PatType};

/// Convert a function to a zero-arg function, and return the args
///
/// Panics if any arg is a receiver (i.e. `self` or a variant)
pub fn strip_args(mut f: ItemFn) -> (ItemFn, Vec<PatType>) {
    let args = replace(&mut f.sig.inputs, Punctuated::new());
    (f, args.into_iter().map(|arg| match arg {
        FnArg::Typed(arg) => arg,
        FnArg::Receiver(_) => panic!("receivers aren't allowed - should be filtered by `validate`"),
    }).collect())
}

#[cfg(test)]
mod tests {
    use quote::ToTokens;
    use syn::parse_str;

    use super::*;

    #[test]
    fn strip_args_works() {
        let f = parse_str("fn foo(i: i32) {}").unwrap();
        let (f, mut types) = strip_args(f);

        assert_eq!(f.to_token_stream().to_string(), "fn foo () { }");

        assert_eq!(types.len(), 1);
        let ty = types.pop().unwrap();
        assert_eq!(ty.to_token_stream().to_string(), "i : i32");
    }

    #[test]
    #[should_panic]
    fn strip_args_panics_with_self() {
        let f = parse_str("fn foo(self) {}").unwrap();
        strip_args(f);
    }
}
