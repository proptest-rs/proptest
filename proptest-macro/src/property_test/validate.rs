use proc_macro2::TokenStream;
use quote::quote_spanned;
use syn::{spanned::Spanned, FnArg, ItemFn};

/// Validate an `ItemFn` for some basic sanity checks
///
/// Many checks are deferred to rustc (e.g. rustc already errors if you make a test function
/// unsafe, so we just transparently pass unsafe through to the generated function and let rustc
/// emit the error)
pub(super) fn validate(f: &ItemFn) -> Result<(), TokenStream> {
    all_args_non_self(f)?;

    Ok(())
}

fn all_args_non_self(f: &ItemFn) -> Result<(), TokenStream> {
    let first_self_arg = f
        .sig
        .inputs
        .iter()
        .find(|arg| matches!(arg, FnArg::Receiver(_)));

    match first_self_arg {
        None => Ok(()),
        Some(arg) => err(arg, "`self` parameters are forbidden"),
    }
}

/// Helper function to generate `compile_error!()` outputs
fn err(span: impl Spanned, s: &str) -> Result<(), TokenStream> {
    Err(quote_spanned! { span.span() => compile_error!(#s) })
}

#[cfg(test)]
mod tests {
    use syn::parse_str;

    use super::*;

    #[test]
    fn validate_fails_with_self_arg() {
        let invalids = [
            "fn foo(self) {}",
            "fn foo(&self) {}",
            "fn foo(&mut self) {}",
            "fn foo(self: Self) {}",
            "fn foo(self: &Self) {}",
            "fn foo(self: &mut Self) {}",
            "fn foo(self: Box<Self>) {}",
            "fn foo(self: Rc<Self>) {}",
            "fn foo(self: Arc<Self>) {}",
        ];

        for invalid in invalids {
            let f: ItemFn = parse_str(invalid).unwrap();
            assert!(validate(&f).is_err());
        }
    }
}
