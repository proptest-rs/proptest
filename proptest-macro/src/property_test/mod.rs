use proc_macro2::TokenStream;
use syn::parse2;

use self::validate::validate;

mod codegen;
mod options;
mod utils;
mod validate;

#[cfg(test)]
mod tests;

/// try to parse an item, or return the error as a token stream
macro_rules! parse {
    ($e:expr) => {
        match parse2($e) {
            Ok(item) => item,
            Err(e) => return e.into_compile_error(),
        }
    };
}

pub fn property_test(attr: TokenStream, item: TokenStream) -> TokenStream {
    let item_fn = parse!(item);
    let options = parse!(attr);

    if let Err(compile_error) = validate(&item_fn) {
        return compile_error;
    }

    codegen::generate(item_fn, options)
}
