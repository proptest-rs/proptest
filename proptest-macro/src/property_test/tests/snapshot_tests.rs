use syn::{parse_str, ItemFn};

use crate::property_test::{codegen, options::Options};

#[test]
fn basic_derive_example() {
    let f: ItemFn =
        parse_str("fn foo(x: i32, y: String) { let x = 1; }").unwrap();
    let tokens = codegen::generate(f, Options::default());
    let file = syn::parse_file(&tokens.to_string()).unwrap();
    let formatted = prettyplease::unparse(&file);

    insta::assert_snapshot!(formatted);
}
