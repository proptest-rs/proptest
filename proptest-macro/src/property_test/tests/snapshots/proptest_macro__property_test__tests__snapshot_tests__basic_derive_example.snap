---
source: proptest-macro/src/property_test/tests/snapshot_tests.rs
expression: formatted
---
#[test]
fn foo() {
    #[derive(Debug)]
    struct FooArgs {
        field0: i32,
        field1: String,
    }
    impl ::proptest::prelude::Arbitrary for FooArgs {
        type Parameters = ();
        type Strategy = ::proptest::strategy::Map<
            ::proptest::arbitrary::StrategyFor<(i32, String)>,
            fn((i32, String)) -> Self,
        >;
        fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
            use ::proptest::strategy::Strategy;
            ::proptest::prelude::any::<(i32, String)>()
                .prop_map(|(field0, field1)| Self { field0, field1 })
        }
    }
    let config = ::proptest::test_runner::Config {
        test_name: Some(concat!(module_path!(), "::", stringify!($test_name))),
        source_file: Some(file!()),
        ..::proptest::test_runner::Config::default()
    };
    let mut runner = ::proptest::test_runner::TestRunner::new(config);
    let result = runner
        .run(
            &::proptest::strategy::Strategy::prop_map(
                ::proptest::prelude::any::<FooArgs>(),
                |values| {
                    ::proptest::sugar::NamedArguments(stringify!(FooArgs), values)
                },
            ),
            |::proptest::sugar::NamedArguments(_, FooArgs { field0: x, field1: y })| {
                let result = {
                    let x = 1;
                };
                let _ = result;
                Ok(())
            },
        );
    match result {
        Ok(()) => {}
        Err(e) => panic!("{}", e),
    }
}
